src/goto-instrument/dump_c.cpp

```diff
@@ -1470,9 +1470,7 @@ void dump_ct::cleanup_expr(exprt &expr)
     }
 
     optionalt<exprt> clean_init;
-    if(
-      ns.follow(bu.type()).id() == ID_union &&
-      bu.source_location().get_function().empty())
+    if(ns.follow(bu.type()).id() == ID_union)
```

Comment:
You're right, the problem is again with `create_function_scope($$);`. This part is also parsing unions
```cpp
        | type_specifier declarator
          post_declarator_attributes_opt
          {
            $2=merge($3, $2);
            
            // the symbol has to be visible during initialization
            init($$, ID_declaration);
            parser_stack($$).type().swap(parser_stack($1));
            PARSER.add_declarator(parser_stack($$), parser_stack($2));
            create_function_scope($$);
          }
          initializer_opt
          assigns_opt
          requires_opt
          ensures_opt
        {
          $$=$4;

          if(parser_stack($6).is_not_nil()) // Capture assigns clause
            parser_stack($$).add(ID_C_spec_assigns).swap(parser_stack($6));

          // Capture code contract
          if(parser_stack($7).is_not_nil())
            parser_stack($$).add(ID_C_spec_requires).swap(parser_stack($7));
          if(parser_stack($8).is_not_nil())
            parser_stack($$).add(ID_C_spec_ensures).swap(parser_stack($8));

          // add the initializer
          if(parser_stack($5).is_not_nil())
          {
            to_ansi_c_declaration(parser_stack($$)).add_initializer(parser_stack($5));
          }

          // Kill the scope that 'function_head' creates.
          PARSER.pop_scope();

          // We are no longer in any function.
          PARSER.set_function(PARSER.current_scope().prefix.substr(0, PARSER.current_scope().prefix.find(':')));
        }
```
So when we `set_function` in `create_function_scope`, we also endup setting for unions. Any suggestion on how to handle this case?

Reply:
Let's discuss how we can refactor this PR. These changes to the C parser are pretty scary, but I don't want to block the entire set of improvements because of this part.